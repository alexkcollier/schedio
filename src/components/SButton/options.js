export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
