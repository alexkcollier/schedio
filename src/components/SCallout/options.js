export const calloutTypes = ['', 'success', 'warning', 'error']
