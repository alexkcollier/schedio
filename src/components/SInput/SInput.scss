.input {
  @include control;
  @include control-with-icons;
}
