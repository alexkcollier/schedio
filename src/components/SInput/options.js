export const allowed = [
  'text',
  'email',
  'number',
  'password',
  'search',
  'tel',
  'url',
  'date',
  'datetime-local',
  'month',
  'time',
  'week',
].sort();
