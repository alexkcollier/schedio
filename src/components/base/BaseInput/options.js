export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
