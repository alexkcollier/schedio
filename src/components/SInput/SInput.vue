<template>
  <SLabel
    :for="id"
    :is-inline="isInline"
    :is-fullwidth="isFullwidth"
    :is-required="required"
    :hide-optional="hideOptional"
    :label="label"
  >
    <div class="input__container">
      <input
        :id="id"
        v-bind="$attrs"
        :class="{
          'input--invalid': isInvalid,
          'input--has-icon-left': !!iconLeft,
          'input--has-icon-right': !!iconRight,
        }"
        :name="computedName"
        :placeholder="placeholder"
        :required="required"
        :type="type"
        class="input"
        v-on="listeners"
      >

      <SIcon
        v-if="iconLeft"
        :icon="iconLeft"
        :color="iconLeftColor"
        :shade="iconLeftShade"
        class="input__icon input__icon--left"
        aria-hidden="true"
      />

      <SIcon
        v-if="iconRight"
        :icon="iconRight"
        :color="iconRightColor"
        :shade="iconRightShade"
        class="input__icon input__icon--right"
        aria-hidden="true"
      />
    </div>
  </SLabel>
</template>

<script>
import { allowed } from './options'
import InputText from '@/mixins/InputText.mixin'
import ControlIcons from '@/mixins/ControlIcons.mixin'

/**
 * TODO:
 * [] - other inputs
 */

export default {
  name: 'SInput',

  mixins: [InputText, ControlIcons],

  props: {
    type: {
      type: String,
      default: 'text',
      validator: (value) => {
        return !value || allowed.includes(value) || console.error(`type must be one of: ${allowed}`)
      },
    },
  },
}
</script>

<style lang="scss">
.input {
  @include control;
  @include control-with-icons;
}
</style>
