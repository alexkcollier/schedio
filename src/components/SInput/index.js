import SInput from './SInput.vue'

export { SInput }

export default { install: Vue => Vue.component(SInput.name, SInput) }
