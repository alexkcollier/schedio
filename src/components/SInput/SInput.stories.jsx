import { SHeading } from '@/components/SHeading'
import { colors as iconColors } from '@/components/SIcon/options'
import { SInput } from '@/components/SInput'
import { allowed } from '@/components/SInput/options'
import PropList from '@/docs/PropList'
import { withAttrsAsProps, withUnboundAttrs } from '@/mixins/stories/form-fields'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import icons from 'feather-icons/dist/icons.json'

const iconList = Object.keys(icons)
const iconProp = side => select(`icon-${side}`, ['', ...iconList], '', 'Optional Props')
const iconColor = side =>
  select(`icon-${side}-color`, ['', ...iconColors], 'green', 'Optional Props')

storiesOf('Components/Inputs.SInput', module)
  .addParameters({ jest: 'SInput' })
  .add('Input', () => {
    return {
      props: {
        props: {
          default: () => ({
            ...withAttrsAsProps({ id: 'base-input', name: 'base-input', label: 'Base Input' }),
            placeholder: text('placeholder', 'Placeholder text', 'Required Props'),
            type: select('type', ['', ...allowed], '', 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props'),
            iconLeft: iconProp('left'),
            iconLeftColor: iconColor('left'),
            iconRight: iconProp('right'),
            iconRightColor: iconColor('right')
          })
        },

        attrs: {
          default: withUnboundAttrs()
        }
      },
      render(h) {
        const { props, attrs } = this.$props
        return (
          <div>
            <SHeading level="1">Input</SHeading>
            <p>Inputs are used to collect user provided text data that should be fairly short.</p>

            <SHeading level="2">Example</SHeading>
            <SInput {...{ props }} {...{ attrs }} />

            <PropList component={SInput} />
          </div>
        )
      }
    }
  })
