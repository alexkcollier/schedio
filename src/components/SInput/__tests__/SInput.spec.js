import { colors as iconColors } from '@/components/SIcon/options'
import { allowed as options } from '@/components/SInput/options'
import SInput from '@/components/SInput/SInput.vue'
import { mount, shallowMount } from '@vue/test-utils'
import icons from 'feather-icons/dist/icons.json'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)
const iconList = Object.keys(icons)

describe('SInput.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { })
  const inputEvent = jest.fn()
  let wrapper

  beforeEach(() => {
    wrapper = mount(SInput, {
      propsData: {
        id: 'input',
        label: 'input',
      },
      listeners: {
        input: inputEvent,
      },
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('requires required props', () => {
    // mounting component without required props
    shallowMount(SInput)
    expect(errorSpy).toHaveBeenCalled()
    expect(errorSpy.mock.calls[0][0]).toContain('[Vue warn]: Missing required prop')
  })

  it('validates the input type', () => {
    const defaultProps = { id: 'input', label: 'input' }
    const randOptionIndex = Math.round(Math.random() * options.length)
    const option = options[randOptionIndex]

    shallowMount(SInput, {
      propsData: {
        ...defaultProps,
        type: option,
      },
    })
    expect(errorSpy).not.toHaveBeenCalled()

    shallowMount(SInput, {
      propsData: {
        ...defaultProps,
        type: 'checkbox',
      },
    })
    expect(errorSpy).toHaveBeenCalled()
  })

  it('can be required', () => {
    wrapper.setProps({ required: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be invalid', () => {
    wrapper.setProps({ isInvalid: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have a placeholder', () => {
    wrapper.setProps({ placeholder: 'some placeholder' })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('accepts user input', () => {
    const input = wrapper.find('input')
    input.element.value = 'text'
    input.trigger('input')
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(input.element.value).toBe('text')
  })

  it('can have an icon on the left', () => {
    wrapper.setProps({ iconLeft: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have an icon on the right', () => {
    wrapper.setProps({ iconRight: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can set each icon color', () => {
    wrapper.setProps({
      iconLeftColor: iconColors[0],
      iconRightColor: iconColors[1],
    })

    expect(errorSpy).not.toHaveBeenCalled()
    expect(wrapper.props('iconLeftColor')).not.toBe(wrapper.props('iconRightColor'))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can hide the optional tag', () => {
    wrapper.setProps({ hideOptional: true, required: false })
    expect(wrapper.text()).not.toContain('Optional')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
