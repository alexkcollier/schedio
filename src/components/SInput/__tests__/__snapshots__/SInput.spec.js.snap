// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SInput.vue can be invalid 1`] = `<slabel-stub label="input" for="input"><input id="input" name="input" placeholder="" required="required" type="text" class="input input--invalid"></slabel-stub>`;

exports[`SInput.vue can be required 1`] = `<slabel-stub label="input" for="input"><input id="input" name="input" placeholder="" required="required" type="text" class="input"></slabel-stub>`;

exports[`SInput.vue can have a placeholder 1`] = `<slabel-stub label="input" for="input"><input id="input" name="input" placeholder="some placeholder" required="required" type="text" class="input"></slabel-stub>`;

exports[`SInput.vue renders correctly 1`] = `<slabel-stub label="input" for="input"><input id="input" name="input" placeholder="" required="required" type="text" class="input"></slabel-stub>`;
